---
layout: page
title: About Me
tags: [about]
comments: false
---

I am a fifth-year PhD student in the [Storey Lab](http://www.genomine.org/) within the [Quantitative and Computational Biology](http://www.princeton.edu/qcbgrad/) program at Princeton University. My interests include bioinformatics (especially genomics), statistics, data analysis, education, and programming in R and Python.

Software
============

* [broom](http://github.com/dgrtwo/broom): an R package for converting messy model outputs to a tidy format, for use with tools such as [dplyr](http://github.com/hadley/dplyr) and [tidyr](http://github.com/hadley/tidyr).
* [subSeq](http://github.com/StoreyLab/subSeq): Subsample reads from an RNA-Seq experiment to determine if it has sufficient depth.
* [ParsePy](https://github.com/dgrtwo/ParsePy): a Python API client for the [Parse](https://parse.com/) mobile backend.
* [gleam](http://github.com/dgrtwo/gleam): Python package for building interactive online visualizations without web programming.
* [OASIS](https://github.com/dgrtwo/OASIS): Python package for annotating insertion sequences in prokaryotic genomes.
* [BarNone](http://github.com/dgrtwo/barnone): Python package to match barcodes in sequencing data to a catalog based on Levenshtein distance.

Publications
============

1. Robinson, D.G. (2015) <a href="http://arxiv.org/abs/1412.3565">broom: An R package for converting statistical analysis objects into tidy data frames</a>. arXiv preprint. arXiv:1412.3565 [stat.CO].
2. Robinson, D.G., Wang, J., and Storey, J.D. (2014) <a href="http://biorxiv.org/content/early/2014/12/30/013342">A nested parallel experiment demonstrates differences in intensity-dependence between RNA-seq and microarrays</a>. biorXiv preprint. doi:10.1101/013342.
3. Robinson, D.G. and Storey, J.D. (2014) <a href="http://bioinformatics.oxfordjournals.org/content/early/2014/09/03/bioinformatics.btu552.abstract?keytype=ref&ijkey=al7iHAgvbsLcdYj">subSeq: Determining appropriate sequencing depth through efficient read subsampling</a>. Bioinformatics, 30 (23): 3424-3426. doi: 10.1093/bioinformatics/btu552.
4. Robinson, D.G., Chen, W., Storey, J.D., and Gresham, D. (2014) <a href="http://www.g3journal.org/content/early/2013/10/30/g3.113.008565.abstract">Design and Analysis of Bar-seq Experiments</a>. G3: Genes/Genomes/Genetics, 4(1), 11-18
5. Robinson, D.G., Lee, M.C. and Marx, C.J. (2012) <a href="http://nar.oxfordjournals.org/content/early/2012/08/14/nar.gks778.full">OASIS: an automated program for global investigation of bacterial and archaeal insertion sequences</a>. Nucleic Acids Research, 10.1093/nar/gks778.
6. Vuilleumier, S., Chistoserdova, L., Lee, M.C.... Robinson, D.G.... (2009) <a href="http://www.plosone.org/article/info:doi/10.1371/journal.pone.0005584">Methylobacterium genome sequences: a reference blueprint to investigate microbial metabolism of C1 compounds from natural and industrial sources.</a> PloS one, 4, e5584.

Links
=====

* I built an online R video course, [Data Analysis and Visualization Using R](/RData/).

* [My CV](/files/DavidRobinsonResume.pdf)

* [My GitHub](https://github.com/dgrtwo)

<a href="http://stackoverflow.com/users/712603/david-robinson">
<img src="http://stackoverflow.com/users/flair/712603.png" width="208" height="58" alt="profile for David Robinson at Stack Overflow, Q&amp;A for professional and enthusiast programmers" title="profile for David Robinson at Stack Overflow, Q&amp;A for professional and enthusiast programmers">
</a>

About This Site
=========

This site is powered by [Jekyll](http://jekyllrb.com/) using the [Minimal Mistakes](http://mademistakes.com/minimal-mistakes/) theme. All blog posts are released under a [Creative Commons Attribution-ShareAlike 4.0 International License](http://creativecommons.org/licenses/by-sa/4.0/).

All blog posts are compiled with [knitr](http://yihui.name/knitr/) [R markdown](http://rmarkdown.rstudio.com/) using [this script](https://github.com/dgrtwo/dgrtwo.github.com/blob/master/_scripts/knitpages.R). You can find the reproducible sources of each blog post [here](https://github.com/dgrtwo/dgrtwo.github.com/tree/master/_R).
